src/wp-includes/meta.php

```diff
@@ -164,10 +167,126 @@ function add_metadata( $meta_type, $object_id, $meta_key, $meta_value, $unique =
 	return $mid;
 }
 
+/**
+ * Adds multiple items of metadata for the specified object.
+ *
+ * This function is more performant than calling `add_metadata()` multiple times because it queries the database only
+ * once and clears the meta cache only once.
+ *
+ * This function will always insert all of the provided metadata even if matching keys already exist. This behaviour
+ * matches that of add_metadata() when its `$unique` parameter is set to false.
+ *
+ * If the insert fails, no metadata will be inserted. It's not possible for a subset of the rows to be inserted.
+ *
+ * Examples:
+ *
+ *     bulk_add_metadata(
+ *         'post',
+ *         $post_id,
+ *         array(
+ *             'meta_key_1' => 'value_1',
+ *             'meta_key_2' => 'value_2',
+ *         )
+ *     );
+ *
+ * For historial reasons both the meta key and the meta value are expected to be "slashed" (slashes escaped) on input.
+ *
+ * @since x.y.z
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @param string              $meta_type   Type of object metadata is for. Accepts 'post', 'comment', 'term', 'user',
+ *                                         or any other object type with an associated meta table.
+ * @param int                 $object_id   ID of the object metadata is for.
+ * @param array<string,mixed> $meta_fields Metadata values keyed by their meta key. Values must be serializable if non-scalar.
+ * @return array<string,int>|false Array of meta IDs keyed by their meta key on success, false on failure.
+ */
+function bulk_add_metadata( string $meta_type, $object_id, array $meta_fields ) {
+	global $wpdb;
+
+	if ( ! $meta_type || ! $meta_fields || ! is_numeric( $object_id ) ) {
+		return false;
+	}
+
+	$object_id = absint( $object_id );
+	if ( ! $object_id ) {
+		return false;
+	}
+
+	$table = _get_meta_table( $meta_type );
+	if ( ! $table ) {
+		return false;
+	}
+
+	$meta_subtype = get_object_subtype( $meta_type, $object_id );
+	$column       = sanitize_key( $meta_type . '_id' );
+	$data         = array();
+	$return       = array();
+
+	foreach ( $meta_fields as $meta_key => $meta_value ) {
+		// expected_slashed ($meta_key)
+		$meta_key   = wp_unslash( $meta_key );
+		$meta_value = wp_unslash( $meta_value );
+		$meta_value = sanitize_meta( $meta_key, $meta_value, $meta_type, $meta_subtype );
+
+		/** This filter is documented in wp-includes/meta.php */
+		$check = apply_filters( "add_{$meta_type}_metadata", null, $object_id, $meta_key, $meta_value, false );
+		if ( null !== $check ) {
+			$return[ $meta_key ] = $check;
+			continue;
+		}
+
+		/** This action is documented in wp-includes/meta.php */
+		do_action( "add_{$meta_type}_meta", $object_id, $meta_key, $meta_value );
+
+		$data[] = array(
+			$object_id,
+			$meta_key,
+			maybe_serialize( $meta_value ),
+		);
+	}
+
+	// If there is no data to save, don't attempt to save it.
+	if ( ! $data ) {
+		return $return;
+	}
+
+	$inserted = $wpdb->insert_multiple(
```

Comment:
If using a `db.php` drop-in, it's possible that `wpdb::insert_multiple()` will not be defined, correct? While ideally a drop-in would extend the `wpdb` class, this isn't guaranteed. In that case, shouldn't this do a `method_exists()` check and then fall back to the non-multiple alternative?

Reply:
I've been looking through previous changes to `wpdb` and it seems that it is assumed that any `db.php` drop-in _does_ extend the `wpdb` class. There haven't been that many changes or additions to its public API in recent years where a `db.php` drop-in would be affected, but here's what I found:

* Core-56484 (r54065) with discussion between Sergey and Marius in the comments about a `db.php` drop-in
* Core-49364 (r53897)
* Core-58584 (r56050)

In all these cases, a change was made that ultimately presumes that a `db.php` drop-in extends the `wpdb` class so all its methods are available. If a `db.php` drop-in author does maintain a class that doesn't extend `wpdb` then they would need to keep its methods in sync with `wpdb` anyway.

@SergeyBiryukov @Clorith What do you think?